### PR TITLE
CORE-487: Add concept of transaction ownership and state

### PR DESCRIPTION
**Pull request by michael-brd**

### CODE REVIEW

File: crypto/BRCryptoPrivate.h [hunk 1]
```diff
@@ -65,7 +65,7 @@ extern "C" {
 
     private_extern BRCryptoHash
     cryptoHashCreateAsGEN (BRGenericHash gen);
-    
+
```
Comment (EBGToo) [inline crypto/BRCryptoPrivate.h]: Oh no, everywhere....

File: crypto/BRCryptoTransfer.c [hunk 2]
```diff
@@ -117,7 +117,7 @@ extern BRCryptoTransfer
 cryptoTransferCreateAsBTC (BRCryptoUnit unit,
                            BRCryptoUnit unitForFee,
                            BRWallet *wid,
-                           BRTransaction *tid) {
+                           OwnershipKept BRTransaction *tid) {
     BRAddressParams  addressParams = BRWalletGetAddressParams (wid);
     BRCryptoTransfer transfer      = cryptoTransferCreateInternal (BLOCK_CHAIN_TYPE_BTC, unit, unitForFee);
     transfer->u.btc.tid = tid;
```
Comment (EBGToo) [inline crypto/BRCryptoTransfer.c]: So this assignment into `transfer` is holding a BRTransaction that is owned (`OwnershipKept`) by the caller.  The subsequent release doesn't call BRTransactionFee() - all consistent provided `tid` doesn't disappear...

Comment (michael-brd) [reply crypto/BRCryptoTransfer.c]: Correct, this functions like ETH now.

File: bitcoin/BRWalletManagerEvent.c [hunk 3]
```diff
@@ -149,19 +149,27 @@ typedef struct {
     struct BREventRecord base;
     BRWalletManager manager;
     int rid;
-    BRTransaction *transaction;
+    uint8_t *transaction;
+    size_t transactionLength;
+    uint64_t timestamp;
+    uint64_t blockHeight;
```
Comment (EBGToo) [inline bitcoin/BRWalletManagerEvent.c]: If the timestamp, blockHeight and length are so easily derived from the transaction, why must they be added to the event and flowed through all the event-y interfaces.

Comment (EBGToo) [reply bitcoin/BRWalletManagerEvent.c]: Oh, because it is no longer a BRTransaction, it is just some bytes.

File: bitcoin/BRWalletManager.c [hunk 4]
```diff
@@ -29,6 +29,98 @@
 
 #define BWM_SLEEP_SECONDS       (1 * 60)                  // 5 minutes
 
+#define DEFAULT_TRANSACTION_CAPACITY     20
+
+struct BRTransactionWithStateStruct {
+    uint8_t isDeleted;
+    BRTransaction *refedTransaction;
+    BRTransaction *ownedTransaction;
+};
```
Comment (EBGToo) [inline bitcoin/BRWalletManager.c]: Nice.

File: bitcoin/BRWalletManager.c [hunk 5]
```diff
@@ -696,18 +824,28 @@ _BRWalletManagerBalanceChanged (void *info,
...
-    fileServiceSave(manager->fileService, fileServiceTypeTransactions, tx);
+    pthread_mutex_lock (&manager->lock);
+    BRTransactionWithState txnWithState = BRWalletManagerFindTransactionByHash (manager, tx->txHash);
+    if (NULL == txnWithState) {
+        // first we've seen it, so it came from the network; add it to our list
+        txnWithState = BRWalletManagerAddRefedTransaction (manager, tx);
+    } else {
+        // this is a transaction we've submitted; set the reference transaction from the wallet
+        txnWithState = BRTransactionWithStateSetRefed(txnWithState, tx);
+    }
```
Comment (EBGToo) [inline bitcoin/BRWalletManager.c]: Nice.

File: bitcoin/BRWalletManager.c [hunk 6]
```diff
@@ -785,35 +928,37 @@ _BRWalletManagerSyncEvent(void * context,
...
-     * `bwmSignalWalletEvent` from `bwmHandleWalletEvent` but that is not possible at
-     * the moment.
+     * For SYNC_MANAGER_TXN_SUBMITTED, we queue them, as all transactions are owned
+     * by this wallet manager.
      */
 
     BRWalletManager bwm = (BRWalletManager) context;
     switch (event.type) {
         case SYNC_MANAGER_SET_BLOCKS: {
+            // filestystem changes are NOT queued; they are acted upon immediately
```
Comment (EBGToo) [inline bitcoin/BRWalletManager.c]: Thanks for that.

File: bitcoin/BRWalletManagerEvent.c (hunk 3 above)
Comment (michael-brd) [reply bitcoin/BRWalletManagerEvent.c]: Correct. Getting closer to a unified BTC/GEN client here...

**Pull request merged**